pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1214,4 +1239,86 @@ private int getMaxNumPartitionsPerInstance(InstancePartitions instancePartitions
       return (numPartitions + numInstancesPerReplicaGroup - 1) / numInstancesPerReplicaGroup;
     }
   }
+
+  /**
+   * Validate the committed low level consumer segments to see if its segment store copy is available. Fix the missing segment store copy by asking servers to upload to segment store.
+   * Since uploading to segment store involves expensive compression step (first tar up the segment and then upload), we don't want to retry the uploading. Segment without segment store copy can still be downloaded from peer servers.
+   * @see <a href="https://cwiki.apache.org/confluence/display/PINOT/By-passing+deep-store+requirement+for+Realtime+segment+completion#BypassingdeepstorerequirementforRealtimesegmentcompletion-Failurecasesandhandling">By-passing deep-store requirement for Realtime segment completion:Failure cases and handling</a>
+   */
+  public void uploadToSegmentStoreIfMissing(TableConfig tableConfig) {
+    Preconditions.checkState(!_isStopping, "Segment manager is stopping");
+
+    String realtimeTableName = tableConfig.getTableName();
+    // Get all the LLC segment ZK metadata for this table
+    List<LLCRealtimeSegmentZKMetadata> segmentZKMetadataList = ZKMetadataProvider.getLLCRealtimeSegmentZKMetadataListForTable(_propertyStore, realtimeTableName);
+
+    // Iterate through llc segments and upload missing segment store copy by following steps:
+    //  1. Ask servers which have online segment replica to upload to segment store. Servers return segment store download url after successful uploading.
+    //  2. Update the llc segment ZK metadata by adding segment store download url.
+    for (LLCRealtimeSegmentZKMetadata segmentZKMetadata : segmentZKMetadataList) {
+      String segmentName = segmentZKMetadata.getSegmentName();
+      // Only fix the committed llc segment without segment store copy
+      if (segmentZKMetadata.getStatus() == Status.DONE && (segmentZKMetadata.getDownloadUrl() == null || segmentZKMetadata.getDownloadUrl().isEmpty() || segmentZKMetadata.getDownloadUrl().equals(CommonConstants.Segment.METADATA_URI_FOR_PEER_DOWNLOAD))) {
+        try {
+          if (!isExceededMaxSegmentCompletionTime(realtimeTableName, segmentName, getCurrentTimeMs())) {
```

Comment:
+1 not needed.

Reply:
My understanding is that this function is used here to prevent the scenario in which the segment completion thread has just changed the ZK metadata status to Status.DONE in "step-1" of segment commit, but meanwhile this periodic task triggers, and mistakes this segment to not have a copy in deep store. This check skips the process for the newly created segments.

Reply:
@sajjad-moradi That's the reason why I put this check here. But after reading: [SplitCommit](https://cwiki.apache.org/confluence/display/PINOT/Consuming+and+Indexing+rows+in+Realtime#ConsumingandIndexingrowsinRealtime-SplitCommit), I think Status.DONE is marked only after uploading to segment store. 

Is it the case? @Jackie-Jiang @mcvsubbu

Reply:
The update of URL and the status to DONE will be atomic. So, if the URL says "peer", then "peer" it is. So, we don't need this check for correctness sake.

That said, it is good to elapse some time after the commit (not the same time as maxSegmentCompletionTime), because it is possible that the segment just completed (as you say), and the servers are still in the process of loading the segment. It is not yet ONLINE in externalview. So, it is good to have the check for performance sake.

I would say, declare a new time limit (even if it is the same value as the other one), and use that. The other one is needed for correction, so please don't re-use it for this purpose.

Reply:
Please, this is not over-complication. If we use the same constant to verify both, and someone changes the performance optimization path without paying attention to the correction part, we have a problem.
Just define another constant to whatever you think will be a good time to wait until the externalview is available.

Reply:
@mcvsubbu What I'm suggesting here is to not check the time at all. If the segment is ONLINE in external view, we upload the segment; if not, we just skip it and the next round should be able to handle that. Adding another value won't guarantee the segment is ONLINE in external view.